# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: third_party/rust/naga/src/span.rs
# Commit: ef5dc3e04e5f
# Full Hash: ef5dc3e04e5f271eea0636ab3a495e95cc912f1d
# Author: Dzmitry Malyshau <dmalyshau@mozilla.com>
# Date: 2021-09-04 09:40:24
# Regressor Bug: 1726626
# File Overlap Count: 3
# Description:
#   Bug 1726626 - Move gfx/wgpu into a 3rd party dependency r=jgilbert,bholley
#   
#   This update makes wgpu a vendored dependency instead of having it in gfx/wgpu.
#   
#   ## Notes
# ==============================================================================

diff --git a/third_party/rust/naga/src/span.rs b/third_party/rust/naga/src/span.rs
new file mode 100644
--- /dev/null
+++ b/third_party/rust/naga/src/span.rs
@@ -0,0 +1,38 @@
+use std::ops::Range;
+
+// A source code span, used for error reporting.
+#[derive(Clone, Debug, PartialEq)]
+pub enum Span {
+    // Span is unknown - no source information.
+    Unknown,
+    // Byte range.
+    ByteRange(Range<usize>),
+}
+
+impl Default for Span {
+    fn default() -> Self {
+        Self::Unknown
+    }
+}
+
+impl Span {
+    pub fn subsume(&mut self, other: &Self) {
+        match *self {
+            Self::Unknown => self.clone_from(other),
+            Self::ByteRange(ref mut self_range) => {
+                if let Self::ByteRange(ref other_range) = *other {
+                    self_range.start = self_range.start.min(other_range.start);
+                    self_range.end = self_range.end.max(other_range.end);
+                }
+            }
+        }
+    }
+
+    pub fn total_span<'a, T: Iterator<Item = &'a Self>>(from: T) -> Self {
+        let mut span: Self = Default::default();
+        for other in from {
+            span.subsume(other);
+        }
+        span
+    }
+}